### PR TITLE
small tweaks, fixes

### PR DESCRIPTION
**Pull request by david-987**
@amcasey simplified the repo a bit and fixed some stuff you pointed out as I was trying to reproduce a new issue am facing in my real repo, no luck though. 
What's happening is am getting results in node_modules when searching for references (some or all):

In my real repo, when searching for `UtiLClass` references, included in the results are the `UtiLClass` and corresponding `index.ts` file inside the `node_modules` directory of the project it's searching in ie sample results:

```
UtiLClass src\packages\business\node_modules\@packages\util\src\core\util.ts
index.ts src\packages\business\node_modules\@packages\util\src\core
....actual results in the business package...
UtiLClass src\packages\user\node_modules\@packages\util\src\core\util.ts
index.ts src\packages\user\node_modules\@packages\util\src\core
....actual results in the user package....
```
Question is why is it even looking in node_modules in the first place when looking for references? Did I miss an exclude somewhere? The excludes are inherited from the `root` package so they become relative - is this an issue?

```
aws-sdk-test/src/packages/business
$ npx tsc --showConfig
npm WARN config global `--global`, `--local` are deprecated. Use `--location=global` instead.
{
    "compilerOptions": {
        "moduleResolution": "node",
        "isolatedModules": true,
        "resolveJsonModule": true,
        "allowSyntheticDefaultImports": true,
        "esModuleInterop": true,
        "skipLibCheck": true,
        "forceConsistentCasingInFileNames": true,
        "experimentalDecorators": true,
        "composite": true,
        "declaration": true,
        "incremental": true,
        "downlevelIteration": true,
        "declarationMap": true,
        "baseUrl": "../../..",
        "paths": {
            "@packages/*": [
                "./src/packages/*/"
            ]
        },
        "module": "commonjs",
        "target": "es2019",
        "allowJs": true,
        "preserveSymlinks": true,
        "outDir": "./dist",
        "rootDir": "./src"
    },
    "references": [
        {
            "path": "../util"
        }
    ],
    "files": [
        "./src/BusinessClass.ts",
        "./src/index.ts"
    ],
    "include": [
        "src/**/*"
    ],
    "exclude": [
        "../../../node_modules",
        "../../../**/node_modules",
        "../../../**/dist",
        "../../../dist"
    ]
}
```

Worth opening an new issue for this? Having trouble reproducing though, everything here seems the same as live.

### CODE REVIEW

**Comment by david-987** [comment]
Also should mention client can only be imported from client, can import from /server and /packages.
Server can import from /packages.
Packages can only import from other packages.

**Comment by david-987** [comment]
@amcasey New odd TS error when trying to compile the server, just pushed the reproduce here

```
src/product/ServerClass.ts(25,9): error TS2322: Type 'import("C:/Users/David/Documents/Coding/aws-sdk-test/src/packages/user/node_modules/@packages/util/dist/core/Optional").default<import("C:/Users/David/Documents/Coding/aws-sdk-test/src/packages/business/dist/BusinessClass").default>' is not assignable to type 'import("C:/Users/David/Documents/Coding/aws-sdk-test/src/packages/util/dist/core/Optional").default<import("C:/Users/David/Documents/Coding/aws-sdk-test/src/packages/business/dist/BusinessClass").default>'.
  Types have separate declarations of a private property 'value'.
src/product/ServerClass.ts(25,9): error TS2322: Type 'import("C:/Users/David/Documents/Coding/aws-sdk-test/src/packages/user/node_modules/@packages/util/dist/core/Optional").default<import("C:/Users/David/Documents/Coding/aws-sdk-test/src/packages/business/dist/BusinessClass").default>' is not assignable to type 'import("C:/Users/David/Documents/Coding/aws-sdk-test/src/packages/util/dist/core/Optional").default<import("C:/Users/David/Documents/Coding/aws-sdk-test/src/packages/business/dist/BusinessClass").default>'.
  Types have separate declarations of a private property 'value'.
```

Seems TS thinks there's 2 versions of Optional - any ideas why?

**Comment by amcasey** [comment]
@david-987 I'm getting a little lost because the issue we're investigating seems to keep changing.  Maybe it would be easier to have a call?

`exclude` tells TS not to _discover_ files in those folders, but they'll still be pulled in if they imported or triple-slash referenced by files that _are_ included.

Why are your packages in `node_modules`?  In your toy repro, the output files go in `dist` folders alongside the source.  If packages refer to each other through `node_modules`, then you probably want to set up your path mappings differently (and they may not even be required, since TS should be able to map back to the input files from the output files).

**Comment by david-987** [comment]
@amcasey thanks for your patience. The original issue I was having seems to have been solved however am getting stuck in different spots while implementing your suggestions. So apologies if the goalpost keeps moving!

I implemented your suggestion of trying to fix paths, since it's possible our redundant path references are allowing for some unnecessary files being created.
So I removed all  `"peerDependencies"` in package.json for each project to get rid of the nested references in `node_modules` for each project.

The issue now, as you can see in `src\packages\user\src\operations\UserClass.ts`, is that an import such as:
`import { UtiLClass, Optional } from '@packages/util/core';` will not work 
`import { UtiLClass, Optional } from '@packages/util';` does still work.

I think this is due to NodeJs style mode resolution which looks for a package.json (finds one for /util but not for /util/core) and then looks for `node_modules`, but because we removed these, these imports stopped working.

Is this something we have to adjust in our root ts config settings file? 
`"paths": { "@packages/*": ["./src/package/*/"] }` - does this have to be tweaked?

We'd like to be able to import from specific folders within the projects without having to make each folder its own project with a package.json

Thanks in advance!

**Comment by amcasey** [comment]
Sorry, I'm still not understanding how the files end up in `node_modules`.  The output directories appear to be `dist` folders adjacent to the input directories?  I'm probably missing something obvious.

>  since it's possible our redundant path references are allowing for some unnecessary files being created

Can you please elaborate?  I'm not sure what you mean.  I wouldn't expect either path mappings or project references to result in extra files being created.

> So I removed all "peerDependencies" in package.json

Are you pulling down copies of packages in your repo from a package repository (versus using the copy in the repo)?  I still feel like I'm missing something obvious.

> `import { UtiLClass, Optional } from '@packages/util/core';` will not work
> `import { UtiLClass, Optional } from '@packages/util';` does still work.

To be honest, I can't recall how path mappings work with deep imports.  I wouldn't be shocked to learn that they don't and you need to be explicit (e.g. create a mapping for `@packages/util/*`).

> Is this something we have to adjust in our root ts config settings file?

It looks like you might have left an uncensored copy of your project name in your example.  Assuming you meant `@package`, yes, it seems possible that you need to create a new entry for each package you wish to deep import from.

**Comment by david-987** [comment]
> Sorry, I'm still not understanding how the files end up in `node_modules`. The output directories appear to be `dist` folders adjacent to the input directories? I'm probably missing something obvious.

@amcasey They end up there as a result of ,
```    "peerDependencies": {
        "@packages/business": "file:../business",
        "@packages/util": "file:../util"
    }
```
in package.json. They're created after a npm install.


> Can you please elaborate? I'm not sure what you mean. I wouldn't expect either path mappings or project references to result in extra files being created.

peerDependencies results in the extra files.

> Are you pulling down copies of packages in your repo from a package repository (versus using the copy in the repo)? I still feel like I'm missing something obvious.

No


> To be honest, I can't recall how path mappings work with deep imports. I wouldn't be shocked to learn that they don't and you need to be explicit (e.g. create a mapping for `@packages/util/*`).

Interesting... seems like a lot of work for larger projects with many levels. 

> > Is this something we have to adjust in our root ts config settings file?
> 
> It looks like you might have left an uncensored copy of your project name in your example. Assuming you meant `@package`, yes, it seems possible that you need to create a new entry for each package you wish to deep import from.

hehe yes. I meant @packages



As a general comment, does this repo not work for you? What seems off about it? (aside from the intentional bad import in UserClass.ts)

**Comment by amcasey** [comment]
> @amcasey They end up there as a result of ,
> 
> ```
>         "@packages/business": "file:../business",
>         "@packages/util": "file:../util"
>     }
> ```
> 
> in package.json. They're created after a npm install.

I fetched this PR commit and ran `npm ci`, but I don't see a `node_modules/@packages`.  Am I missing a step?

> > To be honest, I can't recall how path mappings work with deep imports. I wouldn't be shocked to learn that they don't and you need to be explicit (e.g. create a mapping for `@packages/util/*`).
> 
> Interesting... seems like a lot of work for larger projects with many levels.

I was mistaken, it does handle nesting.  At least on my box, `@packages/util/core` isn't working because core is actually in `@packages/util/dist/core`.  If you want to be able to omit the `dist`, I believe you'll need another path mapping.

> As a general comment, does this repo not work for you? What seems off about it? (aside from the intentional bad import in UserClass.ts)

So, obviously, the most complicated part is handling the imports.  It's hard for me to be specific about how you should do that because I don't know what bundlers, etc, you're using.  Generally speaking, you want to write your imports in a way that will make your bundler/runtime happy and then add path mappings until TS can figure out what you're doing.

It's true that having a second set of tsconfig files for esm is clumsy, but there isn't presently a better way to handle that.  I'm a little surprised you don't have a tsconfig.esm.json in your root, but maybe you have some other way of building each of those projects.

I'm told I was mistaken about `include` and the best practice is to write it explicitly in composite projects.  Sorry for the confusion.

I'll admit that I'm a little curious about why you called your settings file `settings.tsconfig.json` but your esm file `tsconfig.esm.json`. 😋 

TL;DR: I don't see any glaring problems.  If your editor experience (find-all-references is fast for me) is good and your output is correct (I don't know how to gauge this), stick with what you've got.

**Comment by david-987** [comment]
> I fetched this PR commit and ran `npm ci`, but I don't see a `node_modules/@packages`. Am I missing a step?

You have to install the individual packages ie. run `npm install` from `/packages/user`, `/packages/util`, etc. A `packages/user/node_modules/@packages` directory will appear for the user/util project. 
Note I removed `peerDependencies` in this PR, so you need to build from master for this to happen.

> I was mistaken, it does handle nesting. At least on my box, `@packages/util/core` isn't working because core is actually in `@packages/util/dist/core`. If you want to be able to omit the `dist`, I believe you'll need another path mapping.

Where's the best place to define these additional path mappings? Subpath exports in package.json?

> It's true that having a second set of tsconfig files for esm is clumsy, but there isn't presently a better way to handle that. I'm a little surprised you don't have a tsconfig.esm.json in your root, but maybe you have some other way of building each of those projects.

Nothing is built from root, the tsconfig there is more for convenience. 

> I'll admit that I'm a little curious about why you called your settings file `settings.tsconfig.json` but your esm file `tsconfig.esm.json`. 😋

Nobody is perfect, heh

> TL;DR: I don't see any glaring problems. If your editor experience (find-all-references is fast for me) is good and your output is correct (I don't know how to gauge this), stick with what you've got.

I think we're almost there, the issue of having node_modules results pop up in find references is still there in my real repo. 
Removing the `peerDependencies` in package.json cleaned the node_modules up but it now struggles to find nested references to packages. (looking for `/dist` as you mentioned) - wonder if it's possible to make both ts and node happy without peerDependencies set in package.json.

**Comment by amcasey** [comment]
> > I fetched this PR commit and ran `npm ci`, but I don't see a `node_modules/@packages`. Am I missing a step?
> 
> You have to install the individual packages ie. run `npm install` from `/packages/user`, `/packages/util`, etc. A `packages/user/node_modules/@packages` directory will appear for the user/util project. Note I removed `peerDependencies` in this PR, so you need to build from master for this to happen.

That worked.  I've never seen that before, but I don't actually work in monorepos very often.

It sounds like what you want is `--trace-resolution`, which will tell you how each import was resolved (e.g. looked in node_modules, then path mappings, then ...).

> 
> > I was mistaken, it does handle nesting. At least on my box, `@packages/util/core` isn't working because core is actually in `@packages/util/dist/core`. If you want to be able to omit the `dist`, I believe you'll need another path mapping.
> 
> Where's the best place to define these additional path mappings? Subpath exports in package.json?

So, the most convenient place is definitely the root because extending tsconfigs _clobber_ the list, rather than appending to it.  However, there is a cost per path, per project, so a long list shared by a lot of projects can get expensive.  The alternative is to basically mimic the structure of your project references - if you refer to a project, also provide a way to map imports to its files.  Of course, this results in a lot more duplication.  Unfortunately, I'm not aware of a one-size-fits-all solution.

> 
> > It's true that having a second set of tsconfig files for esm is clumsy, but there isn't presently a better way to handle that. I'm a little surprised you don't have a tsconfig.esm.json in your root, but maybe you have some other way of building each of those projects.
> 
> Nothing is built from root, the tsconfig there is more for convenience.

In that case, you might be happier calling the "leaf" tsconfigs "tsconfig.cjs.json" (or whatever), so they'll be ignored by the editor and then just having your root tsconfig (i.e. the one for the editor) include all your files directly and disable emit.  That will make find-all-refs work as you expect, for example.

I feel like there's a layer missing from your repro where lerna or yarn workspaces or something foreach's over your packages?

> > TL;DR: I don't see any glaring problems. If your editor experience (find-all-references is fast for me) is good and your output is correct (I don't know how to gauge this), stick with what you've got.
> 
> I think we're almost there, the issue of having node_modules results pop up in find references is still there in my real repo. Removing the `peerDependencies` in package.json cleaned the node_modules up but it now struggles to find nested references to packages. (looking for `/dist` as you mentioned) - wonder if it's possible to make both ts and node happy without peerDependencies set in package.json.

Again, I don't do this regularly, but it seems strange to me that each package would have its own `node_modules` folder.  I would have thought they'd be hoisted into a root `node_modules`.  Regardless of where they live, if the files in `node_modules` are symlinks to the TS output files (which doesn't appear to be the case now?), then TS should handle them.
